src/main/java/fi/metatavu/metamind/bot/OpenNLPSlotMatcher.java

```diff
@@ -0,0 +1,128 @@
+import com.rabidgremlin.mutters.core.SlotMatcher;
+import com.rabidgremlin.mutters.core.Tokenizer;
+
+import opennlp.tools.namefind.NameFinderME;
+import opennlp.tools.namefind.TokenNameFinderModel;
+import opennlp.tools.util.Span;
+
+/**
+ * THIS IS A COPY OF ORIGINAL CLASS. REMOVE WHEN MUTTERS SUPPORTS LOADING SLOTS FROM URL
+ */
+
+/**
+ * Implements a SlotMatcher that uses OpenNLP's NER framework.
+ * 
+ * @author rabidgremlin
+ *
+ */
+public class OpenNLPSlotMatcher
+    implements SlotMatcher
+{
+  /** Logger. */
+  private Logger log = LoggerFactory.getLogger(OpenNLPSlotMatcher.class);
+
+  /** Map of NER models. */
+  private HashMap<String, TokenNameFinderModel> nerModels = new HashMap<String, TokenNameFinderModel>();
+
+  /** Map of slot models. These share the NER models. */
+  private HashMap<String, TokenNameFinderModel> slotModels = new HashMap<String, TokenNameFinderModel>();
+
+  /** The tokenizer to use. */
+  private Tokenizer tokenizer;
+
+  /**
+   * Constructor. Allows tokenizer to be supplied because NER can use case etc as cues, so may require different
+   * tokenizer than used for intent matching.
+   * 
+   * @param tokenizer The tokenizer to use on an utterance for NER.
+   */
+  public OpenNLPSlotMatcher(Tokenizer tokenizer)
+  {
+    this.tokenizer = tokenizer;
+  }
+
+  /**
+   * This set the NER model to use for a slot.
+   * 
+   * @param slotName The name of the slot. Should match the name of slots on intents added to the matcher.
+   * @param nerModel The file name of the NER model. This file must be on the classpath.
+   */
+  public void addSlotModel(String slotName, URL modelUrl)
+  {
+    TokenNameFinderModel tnfModel = nerModels.get(modelUrl.toExternalForm());
+    if (tnfModel == null)
+    {
+      try
+      {
+        tnfModel = new TokenNameFinderModel(modelUrl);
+      }
+      catch (Exception e)
+      {
+        throw new IllegalArgumentException("Unable to load NER model", e);
+      }
+    }
+
+    slotModels.put(slotName.toLowerCase(), tnfModel);
+  }
+
+  @Override
+  public HashMap<Slot, SlotMatch> match(Context context, Intent intent, String utterance)
+  {
+    String[] utteranceTokens = tokenizer.tokenize(utterance);
+
+    HashMap<Slot, SlotMatch> matchedSlots = new HashMap<Slot, SlotMatch>();
+
+    for (Slot slot : intent.getSlots())
+    {
+      log.debug("Looking for Slot {}", slot.getName());
+
+      TokenNameFinderModel tnfModel = slotModels.get(slot.getName().toLowerCase());
+      if (tnfModel == null)
+      {
+        log.warn("Could not find NER model for slot {}", slot.getName());
+        continue;
+      }
+
+      NameFinderME nameFinder = new NameFinderME(tnfModel);
+      Span[] spans = nameFinder.find(utteranceTokens);
+
+      if (spans.length > 0)
+      {
+        String[] matches = Span.spansToStrings(spans, utteranceTokens);
+
+        log.debug("Matching for {} against {}", slot.getName(), matches);
+
+        // TODO what to do with multi matches?
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)